### ISSUE TITLE
Can't filter by id

### ISSUE THREAD

**Issue by ghost** [fortunejs/fortune-json-api#18]
The filter is ok, unless you try to filter by the id.

I received this exception.

`Unhandled rejection BadRequestError: The field "id" is non-existent.`

**Comment by gr0uch** [comment]
This is expected, I think. Why are you trying to "filter" by ID? If you have the ID of a record, the URL always looks like this:

```
/:type/:id
```

Not sure what the use case is.

**Comment by ghost** [comment]
It's used by Ember to load relationships without making a request for every relationship.

In the adapter we can set the `coalesceFindRequests` option to true, so we can get all in one request:

``` javascript
{
  post: {
    id: 1,
    comments: [1, 2]
  }
}
```

With this option set to false the request will be:

```
GET /comments/1
GET /comments/2
```

And with the option set to true:

```
GET /comments?filter[id]=1,2
```

**Comment by gr0uch** [comment]
Oh, ok. This is an odd case because the JSON API spec says nothing about the `filter` query, but it seems that they rely on non-standard features. So in the special case when IDs are not in the route and `filter[id]` is present, it should actually change the requested IDs.

**Comment by ghost** [comment]
I don’t think that this is a non-standard feature, it’s likely that is an implicit implementation because the filter must filter any column, independent if is primary key or not.

If I figure out how to implement this, i’ll make a pull request.

Iago Sousa
iagosousa@icloud.com

**Comment by gr0uch** [comment]
The spec says absolutely nothing about filtering:

> JSON API is agnostic about the strategies supported by a server. The filter query parameter can be used as the basis for any number of filtering strategies.

http://jsonapi.org/format/#fetching-filtering

So the fact that Ember Data uses the `filter` query at all is non-standard.

**Comment by ghost** [comment]
> The filter query parameter is reserved for filtering data. Servers and clients SHOULD use this key for filtering operations.
> JSON API is agnostic about the strategies supported by a server. The filter query parameter can be used as the basis for any number of filtering strategies.

The problem is that the filter can be applied to any column except ID, that is also a column.

Seems that the script is ignoring the ID as a column.

Iago Sousa
iagosousa@icloud.com

**Comment by gr0uch** [comment]
What I meant was that any sort of filtering is out of the spec, and is ad hoc. For example, if I wanted to filter by aggregate sum of an array of numbers greater than some amount, of course the spec won't bother with that. Any client using the `filter` option is effectively coupled to a particular implementation.

I think that this discussion should belong in JSON API.

**Comment by ghost** [comment]
Oh yes, sorry for the misunderstanding.

Iago Sousa
iagosousa@icloud.com

**Comment by ghost** [comment]
This is a server strategy and does not rely on specification.

**Comment by gr0uch** [comment]
Actually leave it open, I would like to keep track of it since it's useful for Ember Data.
